controllers/nginx/pkg/cmd/controller/nginx.go

```diff
@@ -202,8 +215,43 @@ NGINX master process died (%v): %v
 			conn.Close()
 			time.Sleep(1 * time.Second)
 		}
-		// start a new nginx master process
-		n.start(cmd, done)
+		// restart a new nginx master process if the controller
+		// is not being stopped
+		if n.isStopping {
+			n.stopCh <- struct{}{}
+		} else {
+			n.start(cmd, done)
+		}
+	}
+}
+
+// OnStop gracefully stops the NGINX master process.
+func (n *NGINXController) OnStop() error {
+	glog.Info("stopping NGINX process...")
+	n.isStopping = true
+
+	o, err := exec.Command(n.binary, "-s", "quit", "-c", cfgPath).CombinedOutput()
+	if err != nil {
+		glog.Errorf("%v\n%v", err, string(o))
+	}
+
+	// TODO: If we keep the modified Controller interface
+	// should this be moved as a generic timeout control for the backend to stop
+	// on time ?
+	timer := time.NewTimer(60 * time.Second)
```

Comment:
I think we cannot rely using a timer because if the user sets `terminationGracePeriodSeconds: 5` this will not work

Reply:
You're right. The value should then be configurable.

While running more tests this morning I realized that Nginx has a timeout already kicking in for that scenario:
 [worker_shutdown_timeout](http://nginx.org/en/docs/ngx_core_module.html#worker_shutdown_timeout)]

This might not be needed at all